### PR TITLE
add cropping1d/2d/3d layers

### PR DESCRIPTION
**Pull request by keunwoochoi**
This is a new feature that is being discussed in https://github.com/fchollet/keras/issues/3162. In this repo, https://github.com/keunwoochoi/keras_cropping_layer, I uploaded example code. I tested Cropping2D on theano/tensorflow, I haven't tested it on `Cropping1D/3D` yet. I would like to know if it makes sense to add this feature first before testing 1D/3D cases. 

Some questions are: 
 * Is the API okay? I think there should be some non-symmetric cropping cases. (At least I need it but I am not working on image, so..)
 * In `Cropping1D`, `self.input_spec = [InputSpec(ndim=3)]` exists in `__init__()`, but it seems like this layer needs overriding `build()`as below:
```python
def build(self, input_shape):
        self.input_spec = [InputSpec(shape=input_shape)]
```
 Does it mean that the line in `__init__()` should be removed?

### CODE REVIEW

File: keras/layers/convolutional.py [hunk 1]
```diff
@@ -1346,6 +1346,173 @@ def get_config(self):
...
+            the cropping dimension (axis 1).
+
+    # Input shape
+        3D tensor with shape (samples, axis_to_crop, features)
+
+    # Output shape
+       3D tensor with shape (samples, cropped_axis, features)
+    '''
+
+    def __init__(self, cropping=(1,1), **kwargs):
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Missing space after comma

File: keras/layers/convolutional.py [hunk 2]
```diff
@@ -1346,6 +1346,173 @@ def get_config(self):
         base_config = super(ZeroPadding3D, self).get_config()
         return dict(list(base_config.items()) + list(config.items()))
 
+class Cropping1D(Layer):
+    '''Cropping layer for 1D input (e.g. temporal sequence).
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Better mention that the cropping happens along the time dimension here (axis 1)

File: keras/layers/convolutional.py [hunk 3]
```diff
@@ -1346,6 +1346,173 @@ def get_config(self):
...
+        input_shape = self.input_spec[0].shape
+        return x[:, self.cropping[0][0]:input_shape[1]-self.cropping[0][1], :]
+
+    def get_config(self):
+        config = {'cropping': self.padding}
+        base_config = super(Cropping2D, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+class Cropping2D(Layer):
+    '''Cropping layer for 2D input (e.g. picture).
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Same, mention this is spatial cropping.

File: keras/layers/convolutional.py [hunk 4]
```diff
@@ -1346,6 +1346,173 @@ def get_config(self):
...
+
+    # Arguments
+        padding: tuple of tuple of int (length 2)
```
Comment (fchollet) [inline keras/layers/convolutional.py]: cropping, not padding

File: keras/layers/convolutional.py [hunk 5]
```diff
@@ -1346,6 +1346,173 @@ def get_config(self):
...
+            How many should be trimmed off at the beginning and end of
+            the 2 padding dimensions (axis 3 and 4).
```
Comment (fchollet) [inline keras/layers/convolutional.py]: The indices of the axes depend on the dim ordering. Just say "width and height"

File: keras/layers/convolutional.py [hunk 6]
```diff
@@ -1346,6 +1346,173 @@ def get_config(self):
...
+    # Input shape
+        4D tensor with shape:
+        (samples, depth, first_axis_to_crop, second_axis_to_crop)
+
+    # Output shape
+        4D tensor with shape:
+        (samples, depth, first_cropped_axis, second_cropped_axis)
+    '''
+
+    def __init__(self, cropping=((0,0),(0,0)), dim_ordering='default', **kwargs):
```
Comment (fchollet) [inline keras/layers/convolutional.py]: PEP8: missing commas

File: keras/layers/convolutional.py [hunk 7]
```diff
@@ -1346,6 +1346,173 @@ def get_config(self):
...
+        assert dim_ordering in {'tf', 'th'}, 'dim_ordering must be in {tf, th}'
+        self.dim_ordering = dim_ordering
+        self.input_spec = [InputSpec(ndim=4)]        
+
+    def build(self, input_shape):
+        self.input_spec = [InputSpec(shape=input_shape)]
+
+    def get_output_shape_for(self, input_shape):
+        if self.dim_ordering == 'th':
+
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Unnecessary whitespace

File: keras/layers/convolutional.py [hunk 8]
```diff
@@ -1346,6 +1346,173 @@ def get_config(self):
...
+    def get_config(self):
+        config = {'cropping': self.padding}
+        base_config = super(Cropping2D, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+class Cropping3D(Layer):
+    '''Cropping layer for 2D input (e.g. picture).
+
+    # Arguments
+        padding: tuple of tuple of int (length 2)
```
Comment (fchollet) [inline keras/layers/convolutional.py]: cropping, not padding

File: keras/layers/convolutional.py [hunk 9]
```diff
@@ -1346,6 +1346,173 @@ def get_config(self):
...
+            How many should be trimmed off at the beginning and end of
+            the 2 padding dimensions (axis 3 and 4).
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Same: don't mention specific axes

File: keras/layers/convolutional.py [hunk 10]
```diff
@@ -1346,6 +1346,173 @@ def get_config(self):
...
+        5D tensor with shape:
+        (samples, depth, first_axis_to_crop, second_axis_to_crop, third_axis_to_crop)
+
+    # Output shape
+        5D tensor with shape:
+        (samples, depth, first_cropped_axis, second_cropped_axis, third_cropped_axis)
+    
+    '''
+
+    def __init__(self, cropping=((1,1),(1,1),(1,1)), dim_ordering='default', **kwargs):
```
Comment (fchollet) [inline keras/layers/convolutional.py]: PEP8

File: keras/layers/convolutional.py [hunk 11]
```diff
@@ -1346,6 +1346,173 @@ def get_config(self):
...
+                    dim2,
+                    dim3,
+                    input_shape[4])
+        else:
+            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
+
+    def call(self, x, mask=None):
+        input_shape = self.input_spec[0].shape
+        if self.dim_ordering == 'th':
+            return x[:, :, self.cropping[0][0]:input_shape[2]-self.cropping[0][1], self.cropping[1][0]:input_shape[3]-self.cropping[1][1], self.cropping[2][0]:input_shape[4]-self.cropping[2][1]]
```
Comment (fchollet) [inline keras/layers/convolutional.py]: PEP8

File: keras/layers/convolutional.py [hunk 12]
```diff
@@ -1346,6 +1346,173 @@ def get_config(self):
...
+        elif self.dim_ordering == 'tf':
+            return x[:, self.cropping[0][0]:input_shape[1]-self.cropping[0][1], self.cropping[1][0]:input_shape[2]-self.cropping[1][1], self.cropping[2][0]:input_shape[3]-self.cropping[2][1], :]
```
Comment (fchollet) [inline keras/layers/convolutional.py]: PEP8

**Comment by fchollet** [comment]
It misses tests. LGTM otherwise.

File: keras/layers/convolutional.py (hunk 5 above)
Comment (keunwoochoi) [inline keras/layers/convolutional.py]: Is it 'width and height' or 'height and width'? I though it's the latter.

Comment (fchollet) [inline keras/layers/convolutional.py]: It's width and height. Doesn't matter tbh.

**Comment by keunwoochoi** [comment]
I ran the test and it fails at 
```python

keras/utils/test_utils.py:117:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/keras/layers/test_convolutional.py:154: in test_atrous_conv_2d
    input_shape=(nb_samples, stack_size, nb_row, nb_col))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

layer_cls = <class 'keras.layers.convolutional.AtrousConvolution2D'>
```
...
```python
>       assert expected_output_shape == actual_output_shape
E       assert (2, 2, 6, 2) == (2, 2, 8, 4)
E         At index 2 diff: 6 != 8
E         Full diff:
E         - (2, 2, 6, 2)
E         ?        ^  ^
E         + (2, 2, 8, 4)
E         ?        ^  ^
```
which I don't think is relevant to the part I added. 

There are three FAILs, which are,
```
[gw0] FAILED tests/integration_tests/test_temporal_data_tasks.py::test_temporal_classification <- keras/utils/test_utils.py
tests/integration_tests/test_temporal_data_tasks.py::test_masked_temporal <- keras/utils/test_utils.py

[gw0] FAILED tests/keras/backend/test_backends.py::TestBackend::test_ctc
tests/keras/backend/test_backends.py::TestBackend::test_one_hot

[gw1] FAILED tests/keras/layers/test_convolutional.py::test_atrous_conv_2d <- keras/utils/test_utils.py
tests/keras/layers/test_convolutional.py::test_averagepooling_2d <- keras/utils/test_utils.py
```

**Comment by fchollet** [comment]
Tests are passing on Travis, so the issue is with your version of Theano.

**Comment by keunwoochoi** [comment]
Is it ready now? 
(PS. will group commits next time. 
PPS. the test still fails after `pip install keras --upgrade`, on Mac, py2.7.)

**Comment by fchollet** [comment]
LGTM. But a layer test is still necessary.

**Comment by keunwoochoi** [comment]
Okay. How does it work? Do you want me to add some code on [`test_convolutional.py`](https://github.com/fchollet/keras/blob/2ac6811362aafdbf275c18d634ae6ba78cea406d/tests/keras/layers/test_convolutional.py)?

**Comment by fchollet** [comment]
Basically, yes, add one unit test per new layer in test_convolutional.

File: keras/layers/convolutional.py [hunk 13]
```diff
@@ -1346,6 +1346,191 @@ def get_config(self):
...
+                    input_shape[1] - self.cropping[0][0] - self.cropping[0][1],
+                    input_shape[2] - self.cropping[1][0] - self.cropping[1][1],
+                    input_shape[3])
+        else:
+            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
+
+    def call(self, x, mask=None):
+        input_shape = self.input_spec[0].shape
+        if self.dim_ordering == 'th':
+            return x[:, :, self.cropping[0][0]:input_shape[2]-self.cropping[0][1], self.cropping[1][0]:input_shape[3]-self.cropping[1][1]]
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Line too long, and not PEP8 compliant. Break it down into a few lines.

File: keras/layers/convolutional.py [hunk 14]
```diff
@@ -1346,6 +1346,191 @@ def get_config(self):
...
+        elif self.dim_ordering == 'tf':
+            return x[:, self.cropping[0][0]:input_shape[1]-self.cropping[0][1], self.cropping[1][0]:input_shape[2]-self.cropping[1][1], :]
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Same.

File: keras/layers/convolutional.py (hunk 12 above)
Comment (fchollet) [inline keras/layers/convolutional.py]: Same.

File: tests/keras/layers/test_convolutional.py [hunk 15]
```diff
@@ -444,5 +444,81 @@ def test_upsampling_3d():
...
+        # correctness test
+        layer = convolutional.Cropping3D(
+                        cropping=cropping,
+                        dim_ordering=dim_ordering)
+        layer.set_input(K.variable(input), shape=input.shape)
+
+        out = K.eval(layer.output)
+        # compare with numpy
+        if dim_ordering == 'th':
+            expected_out = input[:, :, cropping[0][0]:-cropping[0][1], cropping[1][0]:-cropping[1][1], cropping[2][0]:-cropping[2][1]]
```
Comment (fchollet) [inline tests/keras/layers/test_convolutional.py]: Same.

File: tests/keras/layers/test_convolutional.py [hunk 16]
```diff
@@ -444,5 +444,81 @@ def test_upsampling_3d():
...
+        else:
+            expected_out = input[:, cropping[0][0]:-cropping[0][1], cropping[1][0]:-cropping[1][1], cropping[2][0]:-cropping[2][1], :]
```
Comment (fchollet) [inline tests/keras/layers/test_convolutional.py]: Same.

**Comment by keunwoochoi** [comment]
Finally, it passes all the tests. Sorry for many mistakes.

File: keras/layers/convolutional.py [hunk 17]
```diff
@@ -1346,6 +1346,205 @@ def get_config(self):
         base_config = super(ZeroPadding3D, self).get_config()
         return dict(list(base_config.items()) + list(config.items()))
 
+class Cropping1D(Layer):
+    '''Cropping layer for 1D input (e.g. temporal sequence).
+    It crops along the time dimension (axis 1).
+
+    # Arguments
+        cropping: tuple of int (length 2)
+            How many should be trimmed off at the beginning and end of
```
Comment (fchollet) [inline keras/layers/convolutional.py]: "how many units" (typo)

File: keras/layers/convolutional.py [hunk 18]
```diff
@@ -1346,6 +1346,205 @@ def get_config(self):
...
+        base_config = super(Cropping1D, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+class Cropping2D(Layer):
+    '''Cropping layer for 2D input (e.g. picture).
+    It crops along spatial dimensions, i.e. width and height.
+
+    # Arguments
+        cropping: tuple of tuple of int (length 2)
+            How many should be trimmed off at the beginning and end of
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Same.

File: keras/layers/convolutional.py [hunk 19]
```diff
@@ -1346,6 +1346,205 @@ def get_config(self):
...
+        (samples, depth, first_axis_to_crop, second_axis_to_crop)
+
+    # Output shape
+        4D tensor with shape:
+        (samples, depth, first_cropped_axis, second_cropped_axis)
+
+    # Examples
+
+    ```python
+        # crop the input image and feature meps
```
Comment (fchollet) [inline keras/layers/convolutional.py]: meps?

File: keras/layers/convolutional.py [hunk 20]
```diff
@@ -1346,6 +1346,205 @@ def get_config(self):
...
+        5D tensor with shape:
+        (samples, depth, first_cropped_axis, second_cropped_axis, third_cropped_axis)
+    
+    '''
+
+    def __init__(self, cropping=((1, 1), (1, 1), (1, 1)), dim_ordering='default', **kwargs):
+        super(Cropping3D, self).__init__(**kwargs)
+        if dim_ordering == 'default':
+            dim_ordering = K.image_dim_ordering()
+        self.cropping = tuple(cropping)
```
Comment (fchollet) [inline keras/layers/convolutional.py]: At this point a check should be done that the `cropping` argument is a tuple of length 3 of tuples of length 2

File: keras/layers/convolutional.py [hunk 21]
```diff
@@ -1346,6 +1346,205 @@ def get_config(self):
...
+
+    ```
+
+    '''
+
+    def __init__(self, cropping=((0, 0), (0, 0)), dim_ordering='default', **kwargs):
+        super(Cropping2D, self).__init__(**kwargs)
+        if dim_ordering == 'default':
+            dim_ordering = K.image_dim_ordering()
+        self.cropping = tuple(cropping)
```
Comment (fchollet) [inline keras/layers/convolutional.py]: At this point a check should be done that the cropping argument is a tuple of length 2 of tuples of length 2

**Comment by fchollet** [comment]
LGTM. Still a few minor touch-ups required, but otherwise it looks good to merge.

**Comment by keunwoochoi** [comment]
Hm....it's bit strange. Test has failed at some strange code.

**Comment by fchollet** [comment]
Cool, LGTM. Thanks!

File: keras/layers/convolutional.py [hunk 22]
```diff
@@ -1346,6 +1346,213 @@ def get_config(self):
...
+        assert len(self.cropping) == 3, 'cropping must be a tuple length of 3'
+        assert len(self.cropping[0]) == 2, 'cropping[0] must be a tuple length of 2'
+        assert len(self.cropping[1]) == 2, 'cropping[1] must be a tuple length of 2'
+        assert len(self.cropping[2]) == 2, 'cropping[2] must be a tuple length of 2'
+        assert dim_ordering in {'tf', 'th'}, 'dim_ordering must be in {tf, th}'
+        self.dim_ordering = dim_ordering
+        self.input_spec = [InputSpec(ndim=4)]
```
Comment (felixlaumon) [inline keras/layers/convolutional.py]: Should be `[InputSpec(ndim=5)]`
